### ISSUE TITLE
Enable team to Train their Own Auto Encoding Models

### ISSUE THREAD

**Issue by keenanjohnson** [Seattle-Aquarium/CCR_image_processing#13]
As we are currently in a somewhat experimental phase for the exact process of what input imagery and parameters to use to train the ML models, it would be helpful if it was easy for anyone on the Seattle aquarium team to kick off a training run. 

Due to the performance constraints of the machines required to train these models in a reasonable timeframe and the budget constraints of the team for now, it's previously been discussed that running training in the cloud would likely be the best option to limit upfront spending while providing access to the large memory and GPU / TPU hardware to accelerate the training. 

This has several components:

1. Create an easy way for any machine to access the training dataset - Complete!
2. Ensure that the data preparation and training scripts are fully documented and tested - Complete!
3. Provision the cloud computer resource (current plan is google cloud VM with either a TPU or GPU) - In progress
4. Set up an access control / login mechanism for the team to access the shared resources to start, stop, and monitor training - Not started
5. Teach everyone to use it - Not started

**Comment by keenanjohnson** [comment]
1. Is now complete thanks to the great work of @reidtosa moving the initial dataset to hugging face (https://huggingface.co/datasets/Seattle-Aquarium/Seattle_Aquarium_benthic_imagery). I believe this is a nice solution as it allows us to create new datasets or subsets of the dataset to train with different combinations using a standard home and way to access the data. 

The training repo supports this now with the merging of the following PRs:

https://github.com/keenanjohnson/underwater-auto-image-encoder/pull/29 

and 

https://github.com/keenanjohnson/underwater-auto-image-encoder/pull/31

**Comment by keenanjohnson** [comment]
I believe number 2 is also complete based on my readme in https://github.com/keenanjohnson/underwater-auto-image-encoder, though always looking to improve it.

**Comment by keenanjohnson** [comment]
Number 3 is in work still. I have initially tried to set up the google TPU but haven't successfully done that. So I have switched over to testing with a nvidia GPU in the google cloud. Will report back here with status.

**Comment by zhrandell** [comment]
Super cool, thanks for all of this, Keenan! We'll try to get it all up and running on the mapping station, and we'll report back run times, etc., such that we can start to get a sense for how useful the mapping station GPU will be for this work.

**Comment by reidtosa** [comment]
I let the machine run after I had left for the day and didn't get a look at that portion :(
I figured my terminal would be waiting for me this morning but it had closed at some point overnight. If it provides any insight, I started the process around 3:40 pm yesterday and my "best_model.pth" file has a creation time of 4:37 pm and modified time of 7:17 pm.

**Comment by keenanjohnson** [comment]
Ah :( at IT issue. 

Yeah that's a good idea to record the text file logs. 

For more context these are what the logs look like below. The important bits are:

Epoch ( 9 /150 ) for example is the iteration count. As I mentioned, we likely need this to be 50+ iterations to start seeing models with good results in my experience. 

The other important thing is the info section after each validation step runs.

2025-11-19 13:57:38,879 - INFO - Val Loss: 0.0446, Val PSNR: 22.96 dB
2025-11-19 13:57:40,251 - INFO - ✓ Saved best model: output/best_model.pth

The Val Loss and Val PSNR are the mathematical representation of how close the model is to the human training output. The lower the loss and higher DB the better. 

```
Epoch 9/150
2025-11-19 13:51:57,450 - INFO - --------------------------------------------------------------------------------
Training: 100%|███████████████████████████████████████████████████| 172/172 [04:23<00:00,  1.53s/it, Loss=0.0436, PSNR=23.03 dB]
Validation: 100%|███████████████████████████████████████████████████| 43/43 [01:17<00:00,  1.80s/it, Loss=0.0471, PSNR=22.42 dB]
2025-11-19 13:57:38,879 - INFO - Train Loss: 0.0455, Train PSNR: 22.72 dB
2025-11-19 13:57:38,879 - INFO - Val Loss: 0.0446, Val PSNR: 22.96 dB
2025-11-19 13:57:40,251 - INFO - ✓ Saved best model: output/best_model.pth
2025-11-19 13:57:40,251 - INFO - 
Epoch 10/150
2025-11-19 13:57:40,251 - INFO - --------------------------------------------------------------------------------
Training: 100%|███████████████████████████████████████████████████| 172/172 [04:23<00:00,  1.53s/it, Loss=0.0441, PSNR=22.95 dB]
Validation: 100%|███████████████████████████████████████████████████| 43/43 [01:17<00:00,  1.81s/it, Loss=0.0458, PSNR=22.73 dB]
2025-11-19 14:03:21,144 - INFO - Train Loss: 0.0443, Train PSNR: 22.95 dB
2025-11-19 14:03:21,144 - INFO - Val Loss: 0.0433, Val PSNR: 23.20 dB
2025-11-19 14:03:22,503 - INFO - ✓ Saved best model: output/best_model.pth
2025-11-19 14:03:23,303 - INFO - ✓ Saved periodic checkpoint: checkpoints/checkpoint_epoch_10.pth
2025-11-19 14:03:23,303 - INFO - 
Epoch 11/150
2025-11-19 14:03:23,303 - INFO - --------------------------------------------------------------------------------
Training: 100%|███████████████████████████████████████████████████| 172/172 [04:23<00:00,  1.53s/it, Loss=0.0443, PSNR=23.02 dB]
Validation: 100%|███████████████████████████████████████████████████| 43/43 [01:17<00:00,  1.80s/it, Loss=0.0447, PSNR=22.90 dB]
2025-11-19 14:09:04,600 - INFO - Train Loss: 0.0436, Train PSNR: 23.08 dB
2025-11-19 14:09:04,600 - INFO - Val Loss: 0.0422, Val PSNR: 23.42 dB
2025-11-19 14:09:05,964 - INFO - ✓ Saved best model: output/best_model.pth
2025-11-19 14:09:05,964 - INFO - 
```

**Comment by reidtosa** [comment]
Here's an update on the state of things now that we're one month in:

_Model Training_
At this point, I've experimented with 30+ models derived from in-progress training sessions, best outputs, and checkpoints -- 14 of which I've kept and documented. We have ten U-Net models and four U-Shape Transformer models at the time of writing (see outputs [here](https://www.dropbox.com/scl/fo/4c5l2lgexeg2obd00etlq/ACEY4P0p0Mv7A_xmAvdH0DA?rlkey=kgjekqhbhdvcll23i11p8i96x&st=ezxqz3fe&dl=0); several models uploaded to [Hugging Face](https://huggingface.co/Seattle-Aquarium)). 
- **U-Net**
We've learned that U-Net models do a decent job of replicating colors, particularly for images with homogenous depth, but occasionally create stitch lines/checkerboard patterning across portions of the image (see [#14](https://github.com/Seattle-Aquarium/CCR_image_processing/issues/14#issuecomment-3582603655)). I've also observed a trend that the further the model trains the more muted and flatter the images become.
Standard training parameters: --image-size 512 --batch-size 8 --epochs 200-300 (best_model typically reached after 100-180 epochs).
- **U-Shape**
U-Shape models do a great job color matching our hand-edited images, but they have a tendency to generate blurry/smudgy spots in select images that severely degrade the level of detail we'd be able to extract locally (see [#15](https://github.com/Seattle-Aquarium/CCR_image_processing/issues/15#issuecomment-3657331956)).
Standard training parameters: --image-size 512 --batch-size 4* --epochs 200-300 (best_model typically reached after 100-250 epochs).
*Mapping station GPU memory exceeded when batch-size set to 8

Additionally, when either model is applied in the UIE, the image outputs are about half the size of our hand-edits. As a result, the outputs retain slightly less detail than the hand-edits. Whether it's enough to affect our classifier model performance is yet to be tested.

_Datasets_
We restructured our training dataset a few times since the initial training. We first trained on all 5,147 hand-edited images in our catalog before filtering our poor-quality images and further separating images into "complex" and "flat" training sets. We also incorporated about 1,000 additional images into our dataset as they became available. The latest models have been trained on the complex and flat sets combined to test the removal of poor-quality images.

_Next Steps_
- We will evaluate how UIE-processed images affect the performance of our classifier model via the framework developed by @m-h-williams (see [#16](https://github.com/Seattle-Aquarium/CCR_image_processing/issues/16)). 
- @keenanjohnson indicated that he may be able to tweak the evaluation metrics used to optimize the model during training to improve model performance.
- Potentially test other model frameworks beyond U-Net and U-Shape.
- Potentially adjust the training framework to exclude image corners and edges, which are most prone to distortion.

**Comment by reidtosa** [comment]
Our attempts to train U-Net and U-Shape models produced results that didn't quite match up to our hand edits as far as quality, color accuracy, and detail retention. However, an initial training with the 3D LUT model have produced results that are leagues above the rest.

Closing this Issue as the team has largely shifted away from our own model training after landing on the 3D LUT model (see [here](https://github.com/Seattle-Aquarium/underwater-auto-image-encoder/issues/79)).
